File: Project_Euler/rootOfNumber.js
/*
In this question we’ll implement a function root that calculates the n’th root of a number. The function takes a nonnegative number x and a positive integer n, and returns the positive n’th root of x within an error of 0.001 (i.e. suppose the real root is y, then the error is: |y-root(x,n)| and must satisfy |y-root(x,n)| < 0.001).

Constraints:
[time limit] 5000ms
[input] float x
0 ≤ x
[input] integer n
0 < n
[output] float

consider space and time complexity
*/

function root(x, n) {
  // your code goes here
  let found = false;
  let top = x;
  let bottom = 0;
  let middle, calculated;
  while (!found) {
    middle = (top + bottom) / 2
    calculated = (Math.pow(middle, n))
    if (Math.abs(x - calculated) <= 0.001) { found = true }
    else {
      x - calculated > 0 ? bottom = middle : top = middle;
    }
  }
  return Math.round(middle * 1000) / 1000
  //return middle.toFixed(3);
}

File: Project_Euler/rootOfNumber.py
def root(x, n):
  high = float(x)
  low = 0.0
  while 1:
    mid = (high + low) / 2
    if abs(x - pow(mid, n)) <= 0.001:
      return round(mid, 3)
    elif x > pow(mid, n):
      low = mid
    else:
      high = mid


print(root(27, 3), 'should be 3') # 3
print(root(7, 3), 'should be 1.93') # 1.93
print(root(9, 2), 'should be 3') # 3
print(root(16, 4), 'should be 2') # 2
print(root(10, 3), 'should be 2.154') # 2.154
print(root(160, 3), 'should be 5.429') # 5.429
